### PR TITLE
Copy to json manually

### PR DESCRIPTION
**Pull request by esoergel**
wait on build
http://manage.dimagi.com/default.asp?162940
@NoahCarnahan @czue @dannyroberts 
It errors on these [two lines](https://github.com/dimagi/commcare-hq/blob/88f70e8c3d813403ab33955331c62e08946a09df/corehq/apps/userreports/tasks.py#L23-L24)

The error we were originally getting was because the corresponding config wasn't found, but I think that was a transient issue.  When I try to run the task manually, I get a `TypeError: cannot deepcopy this pattern object` on a save attempt.  This appears to be a [standing issue](https://github.com/dimagi/jsonobject/issues/115) with jsonobject, as Danny pointed out.  Until a fix for that gets rolled out, this does at least make the config save properly, which lets the task run.

### CODE REVIEW

File: corehq/apps/userreports/models.py [hunk 1]
```diff
@@ -331,7 +350,8 @@ def by_domain(cls, domain):
...
     def by_id(cls, config_id):
-        matching = [ds for ds in cls.all() if ds.get_id == config_id]
-        if not matching:
-            return None
-        return matching[0]
+        for ds in cls.all():
+            if ds.get_id == config_id:
+                return ds
+        raise BadSpecError(_('The data source referenced by this report could '
+                             'not be found.'))
```
Comment (esoergel) [inline corehq/apps/userreports/models.py]: this should fail hard, if I understand its usage correctly.

File: corehq/apps/userreports/tests/test_data_source_config.py [hunk 2]
```diff
@@ -64,6 +65,11 @@ def test_indicators(self):
                 # in the database layer. this should eventually be fixed.
                 self.assertEqual(str(expected_indicators[result.column.id]), result.value)
 
+    def test_serializable_custom_configs(self):
+        for config in CustomDataSourceConfiguration.all():
+            # There are some serialization issues with custom configurations.
+            config.to_json()
```
Comment (esoergel) [inline corehq/apps/userreports/tests/test_data_source_config.py]: this test errors on the current master.

File: corehq/apps/userreports/models.py [hunk 3]
```diff
@@ -222,6 +222,25 @@ def all(cls):
...
+                return {copy(k): copy(v) for k, v in node.items()}
+            elif isinstance(node, list):
+                return [copy(v) for v in node]
+            elif isinstance(node, (basestring, type(None), bool,)):
+                return node
+            else:
+                msg = _("Type {} not supported.  I probably missed something")
+                raise BadSpecError(msg.format(type(node)))
+
+        return copy(self._obj)
```
Comment (esoergel) [inline corehq/apps/userreports/models.py]: the original implementation of this is
```python
def to_json(self):
    self.validate()
    return copy.deepcopy(self._obj)
```

**Comment by dannyroberts** [comment]
This looks good to go from my perspective. Sorry for the jsonobject bug!

**Comment by czue** [comment]
@esoergel sorry i wasn't more available today on this, but I don't think this is the right fix. I think the right thing to do is just not call the two failing lines if it is a `CustomDataSourceConfiguration`. Is there a reason that route didn't work? I'm not exactly sure the best way to do it (would be nice if it didn't get hard-coded right where it's failing but that seems like an okay, and much less obtrusive change than this one)

**Comment by esoergel** [comment]
I've got another branch implementing Cory's suggestion, which seems less impactful, so I'll close this one.  (I can't push to github due to the DDoS at this time though)

**Pull request closed without merging**